tensorflow/core/kernels/map_kernels.h

```diff
@@ -184,6 +185,33 @@ class TensorMapHasKey : public OpKernel {
   }
 };
 
+class TensorMapListKeys : public OpKernel {
+ public:
+  explicit TensorMapListKeys(OpKernelConstruction* c) : OpKernel(c) {
+    OP_REQUIRES_OK(c, c->GetAttr("key_dtype", &key_dtype_));
+  }
+  ~TensorMapListKeys() override {}
+
+  void Compute(OpKernelContext* c) override {
+    const TensorMap* m = nullptr;
+    OP_REQUIRES_OK(c, GetInputMap(c, 0, &m));
+    
+    auto it = m->tensors().begin();
+    size_t sz = m->tensors().size();
+    TensorShape shape = it->first.shape();
```

Comment:
We will need to handle a few more cases here:
1. The map is empty: In this case we can return an empty list.
2. The shapes of all keys do not match: We should raise and error in this case.
3. The type of any key does not match the requested `key_dtype_`. This should be an error as well.

Could you add tests for these?

Reply:
Re: #1, we should return consistent shapes (if we return anything at all). But since we don't know the key shape when the map is empty, we should instead raise an error.

Reply:
Another idea: should we consider returning a TensorList(DT_VARIANT) here? We can still recover the current behavior using ListKeys + TensorListStack. That would defer the design of 1 and 2 to the implementation of TensorListStack. And would also let us fetch a list of keys with heterogenous shapes.

Alternatively we could do what TensorListStack does and add a `key_shape` arg to ListKeys that gets used when the map is empty.

Reply:
I'm thinking we can easily add any of those later, but once we add one it will be very difficult to change. Since neither alternative is ideal, I'd opt to keep our options open for now. Especially since we don't expect to deal much with empty maps in practice.